docs/architecture/adr-037-peer-behaviour.md

```diff
@@ -0,0 +1,145 @@
+# ADR 037: Peer Behaviour Interface
+
+## Changelog
+* 07-03-2019: Initial draft
+
+## Context
+
+The responsibility for signaling and acting upon peer behaviour lacks a single 
+owning component and is heavily coupled with the network stack[<sup>1</sup>](#references). Reactors
+maintain a reference to the `p2p.Switch` which they use to call 
+`switch.StopPeerForError(...)` when a peer misbehaves and 
+`switch.MarkAsGood(...)` when a peer contributes in some meaningful way. 
+While the switch handles `StopPeerForError` internally, the `MarkAsGood` 
+method delegates to another component, `p2p.AddrBook`. This scheme of delegation 
+across Switch obscures the responsibility for handling peer behaviour
+and ties up the reactors in a larger dependency graph when testing.
+
+## Decision
+
+Introduce a `PeerBehaviour` interface and concrete implementations which
+provide methods for reactors to signal peer behaviour without direct
+coupling `p2p.Switch`.  Introduce a ErrPeer to provide 
+concrete reasons for stopping peers.
+
+### Implementation Changes
+
+PeerBehaviour then becomes an interface for signaling peer errors as well
+as for marking peers as `good`.
+
+XXX: It might be better to pass p2p.ID instead of the whole peer but as
+a first draft maintain the underlying implementation as much as
+possible.
+
+```go
+type PeerBehaviour interface {
+    Errored(peer Peer, reason ErrPeer)
+    MarkPeerAsGood(peer Peer)
```

Comment:
The first func now tells what happened (errored), the second one - what to do (mark). Maybe we should rename `MarkPeerAsGood` to `BecomeGood`?

Reply:
Hmm, consistent would be sth like `BehavedWell`, `WasUseful`, `Contributed`? I think the actual name stems from the fact that the switch carries around an addrBook and the method there is `MarkGood`. Sound like we could improve the naming here @melekes 👍 Let's keep this in mind and wait for the implementation though.

Reply:
Good point! Let's ammend the ADR to not have it fall through the cracks during implementation. What about:

``` go
type PeerBehaviour interface {
    Behaved(id ID, reason GoodBehaviourPeer)
    Errored(id ID, reason ErrorBehaviourPeer)
}
```

@brapse & @milosevic Thoughts on this?

Reply:
I really like the `Behaved`, `Errored` dichotomy!

The change from `peer p2p.Peer` -> `id p2p.ID` also further decouples the reactor from the network concern heavy peer interface. 👍 

The introduction of `GoodBehaviourPeer` also seems to complete the symmetry 👍 

Happy to amend:
![image](https://user-images.githubusercontent.com/246/56043921-6d065d80-5d3e-11e9-9c38-e3e8b8d3a04c.png)